internal/charmstore/addentity.go

```diff
@@ -894,3 +894,26 @@ func (e *entityCharm) Actions() *charm.Actions {
 func (e *entityCharm) Revision() int {
 	return e.URL.Revision
 }
+
+type requiredCharm struct {
+	charm   string
+	channel params.Channel
+}
+
+// requiredCharms returns a list of a bundle's charms and the preferred channel
+// for obtaining them.
+func requiredCharms(bd *charm.BundleData) []requiredCharm {
+	list := make([]requiredCharm, 0, len(bd.Applications))
+	for _, app := range bd.Applications {
+		list = append(list, requiredCharm{
+			charm:   app.Charm,
+			channel: params.Channel(app.Channel),
+		})
+	}
+
+	sort.Slice(list, func(a, b int) bool {
```

Comment:
normally we sort things so we have consistent output, I'm not sure that applies here, AFAICS we only use the output for validating the output and the order doesn't really matter there.

Reply:
I have included the sort to match the behavior of bundledata.RequiredCharms (see: https://github.com/juju/charm/blob/v6/bundledata.go#L425) which the store code invoked prior to my changes. Unfortunately, that implementation is oblivious to channels so I had to introduce a tuple type to keep track of {charm, channel} values and call slice.Sort on it.